sendrecv/gst-rust/src/main.rs

```diff
@@ -0,0 +1,480 @@
+extern crate clap;
+extern crate glib;
+extern crate gstreamer as gst;
+extern crate gstreamer_sdp;
+extern crate gstreamer_sdp_sys;
+extern crate gstreamer_webrtc;
+extern crate rand;
+extern crate ws;
+#[macro_use]
+extern crate serde_json;
+
+use gst::prelude::*;
+use gst::{BinExt, ElementExt};
+use rand::Rng;
+use std::sync::{Arc, Mutex};
+
+#[derive(PartialEq, PartialOrd, Eq, Debug)]
+enum AppState {
+    // AppStateUnknown = 0,
+    AppStateErr = 1,
+    ServerConnecting = 1000,
+    ServerConnectionError,
+    ServerConnected,
+    ServerRegistering = 2000,
+    ServerRegisteringError,
+    ServerRegistered,
+    ServerClosed,
+    PeerConnecting = 3000,
+    PeerConnectionError,
+    PeerConnected,
+    PeerCallNegotiating = 4000,
+    PeerCallStarted,
+    PeerCallError,
+}
+#[derive(Debug)]
+enum Error {
+    GlibError(glib::Error),
+    GlibBoolError(glib::BoolError),
+    GstStateChangeError(gst::StateChangeError),
```

Comment:
Instead of this you could also use the `failure` crate and just work with `failure::Error` everywhere (which every error can be converted to, e.g. via the ?-operator). Unless you want to specifically match and handle specific errors, this would simplify the code. See the `gstreamer-rs` examples for some example usage